python/tvm/topi/x86/dense_alter_op.py

```diff
@@ -39,6 +40,17 @@ def _alter_dense_layout(attrs, inputs, tinfos, out_type):
         relay.op.get("nn.dense"), attrs, tinfos, out_type, target
     )
     workload = autotvm.task.get_workload(outs)
+
+    data_type = relay.transform.InferType()(IRModule.from_expr(inputs[0]))["main"].body.checked_type
```

Comment:
Does this work for multi-functions? Something like
```
%1 = %global_var(...);
%2 = nn.dense(%1, %weight);
```

Reply:
As long as `%1` is a `TensorType`, it should be no different. If it cannot be type inferred, something is broken anyway inside that function.

Reply:
What I meant was this line creates a new module with only `%1`. I'm not sure if the function body will be included in this new module or just the function symbol (i.e., `%global_var`). The latter case will crash InferType due to fail to loop up the function body. If this is not the case then we're good.

Reply:
Does such case arise in practice? If so, how that module + the following `dense` is supposed to type check, if we disable alter layout on dense and forget about this PR?

Reply:
Also I want to mention that, normally I would expect that `InferCorrectLayout` would specify the desired layout and layout transform would be inserted automatically. But I got an impression that `InferCorrectLayout` doesn't really work with `AlterOpLayout` (unlike `ConvertLayout`), in that `AlterOpLayout` requires shape information. So we need to 1. get shape and 2. insert layout transform inside this call back. I thought this design is weird but it seems this is the only way.

Reply:
Hmm I'm not sure actually. @anijain2305 @icemelon @yzhliu do you have any idea?

Reply:
normally we get types via `tinfos`, maybe i'm missing something, but why can't we get it in this case? i would not recommend apply infer type in the callback.

Reply:
`tinfos` contains the shape for original inputs https://github.com/apache/tvm/blob/c25b8fae2f29dd9d23aa2487e75450dec5340b75/src/relay/transforms/alter_op_layout.cc#L84-L87

But here, I need to know, "Would `data` input be packed or not (2D or 3D), during AlterOpLayout transform?". So I need types for `new_args` in https://github.com/apache/tvm/blob/c25b8fae2f29dd9d23aa2487e75450dec5340b75/src/relay/transforms/alter_op_layout.cc#L90. But type inference is apparently not applied to `new_arg` there, so the only place to get the type of transformed input is inside topi callback (which feels weird to me).

Reply:
I see. I think it's a design flaw initially, or rather dilemma that layout and shape are closely connected while they are dealt in separate passes.
You mentioned `InferCorrectLayout` doesn't really work with `AlterOpLayout`, I'm not sure this is true, AlterOpLayout shall share the same strategy, https://github.com/apache/tvm/blob/main/src/relay/transforms/transform_layout.h#L336-L372 . I think it is doable as long as 1) input[0]'s layout has been explicitly and correctly inferred and 2) nn.dense's `InferCorrectLayout` takes the input layout into consideration.

Reply:
ok I took a look at this issue again, and I still don't see a way to make use of `InferCorrectLayout` in this context. I added
```
InferCorrectLayoutOutput DenseInferCorrectLayout(const Attrs& attrs,
                                                 const Array<Layout>& new_in_layouts,
                                                 const Array<Layout>& old_in_layouts,
                                                 const Array<tvm::relay::Type>& old_in_types) {
  return InferCorrectLayoutOutput({"NC", "NK"}, {"NC"}, attrs);
}
```
I think this is the only reasonable definition. I was hoping that this would insert "NC8c -> NC" transform on the first input and thus force the input layout  be 2D, but in the callback at https://github.com/apache/tvm/blob/95a32995c9c50b3dfc63eca0ff0d5bcd92b01816/python/tvm/topi/x86/dense_alter_op.py#L31 I still get 3D packed input.

I didn't look at the details of how `LayoutRewriter` and the call back are supposed to interact, but my impression is that in the AlterOpLayout case, we cannot rely on `LayoutRewriter` to insert layout transform, and we need to do EVERYTHING inside the call back. That includes adding necessary layout transform, generating new attributes and contrib op. So I believe `InferCorrectLayout` has no role to play here.

Reply:
Hi @masahi sorry for late reply. Would you mind debug a little bit see if inputs[0]'s layout is correctly inferred? My suspect is inputs[0] becomes unknown layout during the alterlayout, and that's why `InferCorrectLayout` doesn't work.

I implemented `AlterOpLayout` in NNVM at the first place and @merrymercy migrated it to Relay. `InferCorrectLayout` should work with this pass or at least it was my initial intention if I remember correctly.

Reply:
@yzhliu  What does it mean for the layout to be correctly inferred?  For example, I dumped `new_in_layouts` in `DenseInferCorrectLayout` but there is no useful information.

In the test case I added, layout information is gone after avg pooling at https://github.com/apache/tvm/blob/95a32995c9c50b3dfc63eca0ff0d5bcd92b01816/tests/python/relay/test_pass_alter_op_layout.py#L1361. Layout information is not propagated to `dense` and so the only way to tell if layout transform is happening is via its input shape (2d or 3d). I

I'm going to look at how `layout_transform` is inserted for conv2d alter layout.

Reply:
ok found the cause of my confusion and a working solution that aligns well with how conv2d alter layout is implemented. A fix incoming!